File: ml-service/app/monitoring/sentry_config.py
"""
Sentry error tracking configuration for ML Service.

Provides centralized error tracking and reporting for the Flask application.

Requirements: 7.1
"""
import os
from typing import Optional

import sentry_sdk
from sentry_sdk.integrations.flask import FlaskIntegration


def init_sentry(
    dsn: Optional[str] = None,
    environment: Optional[str] = None,
    release: Optional[str] = None,
    traces_sample_rate: float = 0.1,
) -> bool:
    """
    Initialize Sentry error tracking for the ML service.

    Args:
        dsn: Sentry DSN (Data Source Name). If not provided, reads from SENTRY_DSN env var.
        environment: Deployment environment (e.g., "production", "staging").
                    If not provided, reads from SENTRY_ENVIRONMENT env var.
        release: Application release version. If not provided, reads from SENTRY_RELEASE env var.
        traces_sample_rate: Sample rate for performance monitoring (0.0 to 1.0).

    Returns:
        True if Sentry was initialized successfully, False otherwise.
    """
    # Get configuration from environment if not provided
    sentry_dsn = dsn or os.environ.get("SENTRY_DSN")
    sentry_environment = environment or os.environ.get("SENTRY_ENVIRONMENT", "development")
    sentry_release = release or os.environ.get("SENTRY_RELEASE")

    # Skip initialization if no DSN is configured
    if not sentry_dsn:
        return False

    try:
        sentry_sdk.init(
            dsn=sentry_dsn,
            environment=sentry_environment,
            release=sentry_release,
            traces_sample_rate=traces_sample_rate,
            integrations=[
                FlaskIntegration(
                    transaction_style="url",
                ),
            ],
            # Filter out health check endpoints from performance monitoring
            traces_sampler=_traces_sampler,
            # Capture request data for debugging
            send_default_pii=False,
            # Set max breadcrumbs
            max_breadcrumbs=50,
            # Attach stack traces to messages
            attach_stacktrace=True,
            # Before send hook for filtering
            before_send=_before_send,
        )
        return True
    except Exception:
        return False


def _traces_sampler(sampling_context: dict) -> float:
    """
    Custom traces sampler to filter out health check endpoints.

    Args:
        sampling_context: Context containing transaction information.

    Returns:
        Sample rate (0.0 to 1.0) for the transaction.
    """
    # Get the transaction name (URL path)
    transaction_context = sampling_context.get("transaction_context", {})
    name = transaction_context.get("name", "")

    # Don't trace health check or metrics endpoints
    if name in ("/health", "/metrics", "/gpu-status"):
        return 0.0

    # Use default sample rate for other endpoints
    return 0.1


def _before_send(event: dict, hint: dict) -> Optional[dict]:
    """
    Filter events before sending to Sentry.

    Args:
        event: The event to be sent.
        hint: Additional context about the event.

    Returns:
        The event to send, or None to drop it.
    """
    # Filter out certain exception types if needed
    if "exc_info" in hint:
        exc_type, exc_value, _ = hint["exc_info"]
        
        # Don't report client disconnection errors
        if exc_type.__name__ in ("ConnectionResetError", "BrokenPipeError"):
            return None

    return event


def capture_exception(error: Exception, context: Optional[dict] = None) -> None:
    """
    Capture an exception and send it to Sentry.

    Args:
        error: The exception to capture.
        context: Optional additional context to attach to the event.
    """
    with sentry_sdk.push_scope() as scope:
        if context:
            for key, value in context.items():
                scope.set_extra(key, value)
        sentry_sdk.capture_exception(error)


def capture_message(message: str, level: str = "info", context: Optional[dict] = None) -> None:
    """
    Capture a message and send it to Sentry.

    Args:
        message: The message to capture.
        level: Log level ("debug", "info", "warning", "error", "fatal").
        context: Optional additional context to attach to the event.
    """
    with sentry_sdk.push_scope() as scope:
        if context:
            for key, value in context.items():
                scope.set_extra(key, value)
        sentry_sdk.capture_message(message, level=level)


def set_user(user_id: Optional[str] = None, ip_address: Optional[str] = None) -> None:
    """
    Set user context for Sentry events.

    Args:
        user_id: Optional user identifier (anonymized).
        ip_address: Optional IP address (will be anonymized by Sentry).
    """
    sentry_sdk.set_user({
        "id": user_id,
        "ip_address": ip_address,
    })


def add_breadcrumb(
    message: str,
    category: str = "custom",
    level: str = "info",
    data: Optional[dict] = None,
) -> None:
    """
    Add a breadcrumb for debugging context.

    Args:
        message: Description of the breadcrumb.
        category: Category for grouping breadcrumbs.
        level: Log level for the breadcrumb.
        data: Optional additional data to attach.
    """
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data or {},
    )

File: credibility-analyzer/src/services/sentry/sentryConfig.ts
/**
 * Sentry Configuration for React Frontend
 * Provides error tracking and performance monitoring
 * Requirements: 7.4
 */

import * as Sentry from '@sentry/react';

/**
 * Configuration options for Sentry initialization
 */
export interface SentryConfig {
  dsn: string;
  environment: string;
  release?: string;
  tracesSampleRate?: number;
  replaysSessionSampleRate?: number;
  replaysOnErrorSampleRate?: number;
  enabled?: boolean;
}

// Track initialization state
let sentryInitialized = false;

/**
 * Gets the default Sentry configuration from environment variables
 * @returns SentryConfig with values from Vite environment variables
 */
export function getDefaultSentryConfig(): SentryConfig {
  return {
    dsn: import.meta.env.VITE_SENTRY_DSN || '',
    environment: import.meta.env.VITE_APP_ENV || 'development',
    release: import.meta.env.VITE_APP_VERSION,
    tracesSampleRate: parseFloat(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE || '0.1'),
    replaysSessionSampleRate: parseFloat(
      import.meta.env.VITE_SENTRY_REPLAYS_SESSION_SAMPLE_RATE || '0.1'
    ),
    replaysOnErrorSampleRate: parseFloat(
      import.meta.env.VITE_SENTRY_REPLAYS_ON_ERROR_SAMPLE_RATE || '1.0'
    ),
    enabled: import.meta.env.VITE_SENTRY_ENABLED === 'true',
  };
}

/**
 * Checks if Sentry is currently initialized
 * @returns true if Sentry has been initialized
 */
export function isSentryInitialized(): boolean {
  return sentryInitialized;
}


/**
 * Initializes Sentry for the React application
 * Must be called before the React app renders
 * @param config Optional configuration (uses defaults if not provided)
 * Requirements: 7.4
 */
export function initSentry(config?: Partial<SentryConfig>): void {
  const cfg = { ...getDefaultSentryConfig(), ...config };

  // Skip initialization if disabled or no DSN
  if (!cfg.enabled || !cfg.dsn) {
    sentryInitialized = false;
    return;
  }

  Sentry.init({
    dsn: cfg.dsn,
    environment: cfg.environment,
    release: cfg.release,
    
    // Performance monitoring
    tracesSampleRate: cfg.tracesSampleRate ?? 0.1,
    
    // Session replay for debugging
    replaysSessionSampleRate: cfg.replaysSessionSampleRate ?? 0.1,
    replaysOnErrorSampleRate: cfg.replaysOnErrorSampleRate ?? 1.0,
    
    integrations: [
      // Browser tracing for performance monitoring
      Sentry.browserTracingIntegration(),
      // Session replay for debugging user sessions
      Sentry.replayIntegration({
        // Mask all text content for privacy
        maskAllText: true,
        // Block all media for privacy
        blockAllMedia: true,
      }),
    ],
    
    // Filter out known non-critical errors
    beforeSend: (event) => {
      // Filter out AbortError (user cancelled requests)
      if (event.exception?.values?.[0]?.type === 'AbortError') {
        return null;
      }
      // Filter out network errors that are expected
      if (event.exception?.values?.[0]?.type === 'NetworkError') {
        return null;
      }
      return event;
    },
  });

  sentryInitialized = true;
}

/**
 * Creates a Sentry error boundary wrapper component
 * Use this to wrap components that should report errors to Sentry
 * @returns Sentry ErrorBoundary component
 * Requirements: 7.4
 */
export const SentryErrorBoundary = Sentry.ErrorBoundary;

/**
 * Captures an exception with optional context
 * Sends the error to Sentry with additional metadata
 * @param error The error to capture
 * @param context Optional additional context to attach
 * Requirements: 7.4
 */
export function captureException(error: Error, context?: Record<string, unknown>): void {
  if (!sentryInitialized) {
    // Log to console if Sentry is not initialized
    console.error('Error captured (Sentry not initialized):', error, context);
    return;
  }

  Sentry.withScope((scope) => {
    if (context) {
      scope.setExtras(context);
    }
    Sentry.captureException(error);
  });
}

/**
 * Captures a message with optional context
 * @param message The message to capture
 * @param level The severity level
 * @param context Optional additional context
 */
export function captureMessage(
  message: string,
  level: 'fatal' | 'error' | 'warning' | 'info' | 'debug' = 'info',
  context?: Record<string, unknown>
): void {
  if (!sentryInitialized) {
    console.log(`Message captured (Sentry not initialized): [${level}] ${message}`, context);
    return;
  }

  Sentry.withScope((scope) => {
    if (context) {
      scope.setExtras(context);
    }
    Sentry.captureMessage(message, level);
  });
}

/**
 * Sets a tag on the current Sentry scope
 * @param key Tag key
 * @param value Tag value
 */
export function setTag(key: string, value: string): void {
  if (!sentryInitialized) {
    return;
  }
  Sentry.setTag(key, value);
}

/**
 * Sets user information on the current Sentry scope
 * Only use with non-PII identifiers
 * @param user User information (should be anonymized)
 */
export function setUser(user: { id?: string; segment?: string }): void {
  if (!sentryInitialized) {
    return;
  }
  Sentry.setUser(user);
}

/**
 * Adds a breadcrumb to the current Sentry scope
 * Breadcrumbs help trace the path leading to an error
 * @param breadcrumb Breadcrumb data
 */
export function addBreadcrumb(breadcrumb: {
  category?: string;
  message?: string;
  level?: 'fatal' | 'error' | 'warning' | 'info' | 'debug';
  data?: Record<string, unknown>;
}): void {
  if (!sentryInitialized) {
    return;
  }
  Sentry.addBreadcrumb(breadcrumb);
}

/**
 * Resets Sentry initialization state (for testing)
 */
export function resetSentry(): void {
  sentryInitialized = false;
}
